sixtyfps_runtime/rendering_backends/qt/qt_window.rs

```diff
@@ -402,16 +402,65 @@ impl ItemRenderer for QtItemRenderer<'_> {
             TextWrap::no_wrap => 0,
             TextWrap::word_wrap => key_generated::Qt_TextFlag_TextWordWrap,
         };
-        let elide = text.overflow() == TextOverflow::elide && text.wrap() == TextWrap::no_wrap;
+        let elide = text.overflow() == TextOverflow::elide;
         let painter: &mut QPainter = &mut *self.painter;
-        cpp! { unsafe [painter as "QPainter*", rect as "QRectF", fill_brush as "QBrush", string as "QString", flags as "int", font as "QFont", elide as "bool"] {
+        cpp! { unsafe [painter as "QPainter*", rect as "QRectF", fill_brush as "QBrush", mut string as "QString", flags as "int", font as "QFont", elide as "bool"] {
             painter->setFont(font);
             painter->setPen(QPen(fill_brush, 0));
             painter->setBrush(Qt::NoBrush);
             if (!elide) {
                 painter->drawText(rect, flags, string);
+            } else if (!(flags & Qt::TextWordWrap)) {
+                QString elided;
+                QFontMetrics fm(font);
+                while (!string.isEmpty()) {
+                    int pos = string.indexOf('\n');
+                    if (pos < 0) {
+                        elided += fm.elidedText(string, Qt::ElideRight, rect.width());
+                        break;
+                    }
+                    QString line = string.left(pos);
+                    elided += fm.elidedText(line, Qt::ElideRight, rect.width());
+                    elided += '\n';
+                    string = string.mid(pos + 1);
+                }
+                painter->drawText(rect, flags, elided);
             } else {
-                auto elided = QFontMetrics(font).elidedText(string, Qt::ElideRight, rect.width());
+                // elide and word wrap: we need to add the elipsis manually on the last line
+                QString elided = string;
+                QFontMetrics fm(font);
+                QTextLayout layout(string, font);
+                QTextOption options;
+                options.setWrapMode(QTextOption::WordWrap);
+                layout.setTextOption(options);
+                layout.setCacheEnabled(true);
+                layout.beginLayout();
+                int leading = fm.leading();
+                qreal height = 0;
+                int last_line_begin = 0, last_line_size = 0;
+                while (true) {
+                    auto line = layout.createLine();
+                    if (!line.isValid()) {
+                        last_line_begin = string.size();
+                        break;
+                    }
+                    line.setLineWidth(rect.width());
+                    auto lf = QStringView(string).mid(line.textStart(), line.textLength()).indexOf('\n');
+                    if (lf >= 0) {
+                        line.setNumColumns(lf);
```

Comment:
setNumColumns will do the line layout again, it would be more efficient to either call `setLineWidth` or `setNumColumns`.

I'm curious though: Why is the call to `setNumColumns` needed? Is it because you observed QTextLayout not respecting `\n`? If that is why, then I think it would be easier to upfront convert `\n` and `\r\n` to `QChar::LineSeparator` and just call `setLineWidth()` here.

Reply:
Yes, \n was not considered as a line separator. I'll try to do the change you suggest